### PR TITLE
Vibrate on Android when `EditableText` changes selection by dragging handles

### LINKED ISSUE #104639: `TextField` should vibrate when moving the cursor on Android
## Description
When moving the cursor in a native Android text field, the phone vibrates of every character move. I tried to show this in the video. You need to make sure the sounds really loud to hear the vibration.

https://user-images.githubusercontent.com/24459435/170326150-cc8aebd0-fab0-4cad-8862-01cf9e01ffd1.MOV

This is also a setting which can be turned off in the system settings. To reproduce the behavior, make sure you enabled the vibrations.

telegram-cloud-photo-size-2-5219966324162279774-y

Phone: I used a Google Pixel 4a with Android 12.
Flutter Version: `3.1.0-0.0.pre.904`

<details>
<summary>My code</summary>

```dart
import 'package:flutter/material.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      home: Scaffold(
        body: Center(
          child: TextField(),
        ),
      ),
    );
  }
}
```

<details>

### LINKED ISSUE #104640: `TextField` should vibrate when changing the selection on Android
## Description
When changing the selection in a native Android text field, the phone vibrates of every character move. I tried to show this in the video. You need to make sure the sounds really loud to hear the vibration.

https://user-images.githubusercontent.com/24459435/170326834-a4f70e4b-b849-461f-bba6-cdfff2ed080d.MOV

This is also a setting which can be turned off in the system settings. To reproduce the behavior, make sure you enabled the vibrations.

telegram-cloud-photo-size-2-5219966324162279774-y

Phone: I used a Google Pixel 4a with Android 12.
Flutter Version: `3.1.0-0.0.pre.904`

<details>
<summary>My code</summary>

```dart
import 'package:flutter/material.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      home: Scaffold(
        body: Center(
          child: TextField(),
        ),
      ),
    );
  }
}
```

</details>

This ticket is also related to #104551

### PR DESCRIPTION
**Pull request by markusaksli-nc**
Trigger `HapticFeedback.selectionClick()` when text selection changes by dragging handles on Android. Ideally this should fix https://github.com/flutter/flutter/issues/104639 as well but it seems to be a separate case.

Fixes https://github.com/flutter/flutter/issues/104640

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by goderbauer** [comment]
(triage): @justinmc or @LongCatIsLooong could you take a look at this one?

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 1]
```diff
@@ -2551,6 +2551,21 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
       _stopCursorTimer(resetCharTicks: false);
       _startCursorTimer();
     }
+
+    // Trigger haptic feedback when handle drag changes selection on Android.
+    switch (defaultTargetPlatform) {
+      case TargetPlatform.android:
+        if (!selection.isCollapsed && cause == SelectionChangedCause.drag) {
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/widgets/editable_text.dart:2558]: Could you explain why the feedback should only be triggered when the selection isn't collapsed?

**Review by LongCatIsLooong — COMMENTED** [review]
Would it be better if this is moved to 
 https://github.com/flutter/flutter/blob/017fc6c907eada9204ccc9885462f9efe9625bcb/packages/flutter/lib/src/material/text_field.dart#L61

File: packages/flutter/lib/src/widgets/editable_text.dart (hunk 1 above)
Comment (markusaksli-nc) [reply packages/flutter/lib/src/widgets/editable_text.dart:2558]: When the selection is collapsed there are some other feedbacks already (for example long-press word selection). This can result in double feedback in some cases.

Comment (markusaksli-nc) [reply packages/flutter/lib/src/widgets/editable_text.dart:2558]: The long-press start feedback for material `TextField` is at https://github.com/flutter/flutter/blob/fd54ddf5cf4ccc920139cb1e4f3beef7a199cd16/packages/flutter/lib/src/material/text_field.dart#L109

However, the feedback for when expanded selection adds or removes a character (but not when the selection is expanded) can't really be handled by the gesture detector callbacks in `_TextFieldSelectionGestureDetectorBuilder`. The gestures that update the selection incrementally are `onSingleLongTapMoveUpdate` but also text selection overlay handle drag. These both trigger `_handleSelectionChanged` in `EditableTextState` when the selected position updates the selection.

This is why it made sense to me to do it where it can be implemented in the simplest way without accessing the `EditableText` state and comparing selection against previous values in multiple places.

Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/editable_text.dart:2558]: Performing such checks doesn't seem to be particularly hard? If possible the behavior shouldn't be hard-coded into `EditableText`.

File: packages/flutter/lib/src/material/text_field.dart [hunk 2]
```diff
@@ -79,6 +83,13 @@ class _TextFieldSelectionGestureDetectorBuilder extends TextSelectionGestureDete
           );
           break;
       }
+      if (_state._effectiveController.selection != _lastSelection) {
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/material/text_field.dart:86]: The `HapticFeedback` should only trigger if  `onSingleLongTapMoveUpdate` changes the selection I think? So I don't think you need the 2 extra instance variables you can just get the selection at the start of this method and check if it has changed?

Comment (markusaksli-nc) [reply packages/flutter/lib/src/material/text_field.dart]: I thought this didn't work when I tried it but turns out it actually does

File: packages/flutter/lib/src/widgets/editable_text.dart (hunk 1 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/editable_text.dart]: @chunhtai do you by chance have the link to the android code that does that? Also I wonder if the same interaction applies to selectable area

Comment (chunhtai) [reply packages/flutter/lib/src/widgets/editable_text.dart]: no i don't have the link to the code, I can look for it though. and yes this applies to selectableregion/selectionarea as well.

File: packages/flutter/lib/src/widgets/text_selection.dart [hunk 3]
```diff
@@ -537,13 +537,20 @@ class TextSelectionOverlay {
...
+  TextSelection? _lastSelection;
+
   void _handleSelectionHandleChanged(TextSelection newSelection, {required bool isEnd}) {
     final TextPosition textPosition = isEnd ? newSelection.extent : newSelection.base;
     selectionDelegate.userUpdateTextEditingValue(
       _value.copyWith(selection: newSelection),
       SelectionChangedCause.drag,
     );
     selectionDelegate.bringIntoView(textPosition);
+
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/widgets/text_selection.dart:549]: Are these changes still needed?

File: packages/flutter/lib/src/material/text_field.dart [hunk 4]
```diff
@@ -79,6 +80,9 @@ class _TextFieldSelectionGestureDetectorBuilder extends TextSelectionGestureDete
           );
           break;
       }
+      if (_state._effectiveController.selection != lastSelection) {
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/material/text_field.dart:83]: nit: `TextSelection` is directional so `(1, 2] != (2, 1]`, despite the 2 having exactly the same visually highlighted region. Haptic feedback probably isn't needed when the start and the end swap.

File: packages/flutter/lib/src/widgets/editable_text.dart (hunk 1 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/editable_text.dart]: Yeah chromium implemented that for selection changes: https://source.chromium.org/chromium/chromium/src/+/main:content/public/android/java/src/org/chromium/content/browser/selection/SelectionPopupControllerImpl.java;bpv=1;bpt=1?q=TEXT_HANDLE_MOVE&ss=chromium

**Comment by markusaksli-nc** [comment]
Should these changes still be tied to material `TextField` then, or should I move them back to `EditableText`?

**Comment by LongCatIsLooong** [comment]
It would be nice to decouple this from `EditableText`.

**Comment by goderbauer** [comment]
@markusaksli-nc (triage): Do you still have plans to follow up on the feedback given above?

File: packages/flutter/lib/src/widgets/text_selection.dart (hunk 3 above)
Comment (markusaksli-nc) [reply packages/flutter/lib/src/widgets/text_selection.dart:770]: these changes are needed to trigger feedback when selection changes due to either handle being dragged, I'm not sure how I would re-implement this in `SelectionOverlay`.

**Comment by markusaksli-nc** [comment]
I rebased and tried to fix the nit, but it seemed to trigger feedback every frame. Overall I'm not sure what feedback to address here or what the goal is. Why would the behavior for text selection feedback be only implemented in the material text field?

I can look at what would be necessary to expose the android setting from engine, but I'm unfamiliar with the process.

**Comment by goderbauer** [comment]
(triage): @LongCatIsLooong @markusaksli-nc What is the status on this one?

**Comment by markusaksli-nc** [comment]
Can't work on this atm so closing
